crates/pallet-subspace/src/lib.rs

```diff
@@ -1464,12 +1471,15 @@ fn check_vote<T: Config>(
         parent_vote_verification_data
     };
 
-    let pieces_in_segment = vote_verification_data.recorded_history_segment_size
-        / vote_verification_data.record_size
-        * 2;
-    let segment_index = solution.piece_index / u64::from(pieces_in_segment);
-    let position = u32::try_from(solution.piece_index % u64::from(pieces_in_segment))
+    let sector_id = SectorId::new(&(&solution.public_key).into(), solution.sector_index);
+
+    let piece_index = sector_id
+        .derive_piece_index(solution.piece_offset, solution.total_pieces)
+        .map_err(|()| CheckVoteError::TotalNumberOfPiecesCantBeZero)?;
+    let pieces_in_segment = vote_verification_data.pieces_in_segment;
+    let position = u32::try_from(piece_index % u64::from(pieces_in_segment))
         .expect("Position within segment always fits into u32; qed");
+    let segment_index: SegmentIndex = piece_index / SegmentIndex::from(pieces_in_segment);
```

Comment:
I also see a similar pattern in `crates/sc-consensus-subspace/src/slot_worker.rs` line 212 ~ 228, basically, sector_id, piece_index and position are all derived from the solution, it could be helpful for me if we can add some helper function to hide these numeric operations if possible.

Reply:
Indeed. There are a few places where things related to new consensus are derived, I plan to extract them into helper functions a bit later. I just didn't add an explicit TODO in this case. Thanks for noticing though!